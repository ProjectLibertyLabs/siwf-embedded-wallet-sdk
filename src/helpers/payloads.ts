import {
  getEip712BrowserRequestAddProvider,
  getEip712BrowserRequestClaimHandlePayload,
  getEip712BrowserRequestItemizedSignaturePayloadV2,
  getEip712BrowserRequestSiwfSignedRequestPayload,
  ItemizedAction,
} from "@frequency-chain/ethereum-utils";
import { EIP712Document, SignatureFn } from "../types";
import {
  AddProviderPayloadArguments,
  ClaimHandlePayloadArguments,
  CreateSignedLogInPayloadArguments,
  ItemActionsPayloadArguments,
  SiwfResponsePayloadAddProvider,
  SiwfResponsePayloadClaimHandle,
  SiwfResponsePayloadItemActions,
} from "src/siwf-types";
import { isHexString } from "./utils";
import { mockCaip122 } from "../signature-requests";

export async function createSignedAddProviderPayload(
  userAddress: string,
  signatureFn: SignatureFn,
  payloadArguments: AddProviderPayloadArguments,
  extrinsic:
    | "createSponsoredAccountWithDelegation"
    | "grantDelegation" = "createSponsoredAccountWithDelegation",
): Promise<SiwfResponsePayloadAddProvider> {
  const addProviderEip712 = getEip712BrowserRequestAddProvider(
    payloadArguments.authorizedMsaId,
    payloadArguments.schemaIds,
    payloadArguments.expiration,
  ) as EIP712Document;
  const signature = await signatureFn({
    method: "eth_signTypedData_v4",
    params: [userAddress, addProviderEip712],
  });

  return {
    signature: {
      algo: "SECP256K1",
      encoding: "base16",
      encodedValue: signature,
    },
    endpoint: {
      pallet: "msa",
      extrinsic,
    },
    type: "addProvider",
    payload: payloadArguments,
  };
}

export async function createSignedClaimHandlePayload(
  userAddress: string,
  signatureFn: SignatureFn,
  payloadArguments: ClaimHandlePayloadArguments,
): Promise<SiwfResponsePayloadClaimHandle> {
  const claimHandleEip712 = getEip712BrowserRequestClaimHandlePayload(
    payloadArguments.baseHandle,
    payloadArguments.expiration,
  ) as EIP712Document;
  const signature = await signatureFn({
    method: "eth_signTypedData_v4",
    params: [userAddress, claimHandleEip712],
  });

  return {
    signature: {
      algo: "SECP256K1",
      encoding: "base16",
      encodedValue: signature,
    },
    endpoint: {
      pallet: "handles",
      extrinsic: "claimHandle",
    },
    type: "claimHandle",
    payload: payloadArguments,
  };
}
export async function createSignedGraphKeyPayload(
  userAddress: string,
  signatureFn: SignatureFn,
  payloadArguments: ItemActionsPayloadArguments,
): Promise<SiwfResponsePayloadItemActions> {
  const actions: ItemizedAction[] = payloadArguments.actions.map(
    ({ payloadHex }) => {
      if (isHexString(payloadHex)) {
        return {
          actionType: "Add",
          data: payloadHex,
          index: 0,
        };
      } else throw new Error(`Expected HexString: ${payloadHex}`);
    },
  );
  const addItemsEip712 = getEip712BrowserRequestItemizedSignaturePayloadV2(
    payloadArguments.schemaId,
    payloadArguments.targetHash,
    payloadArguments.expiration,
    actions,
  ) as EIP712Document;
  const signature = await signatureFn({
    method: "eth_signTypedData_v4",
    params: [userAddress, addItemsEip712],
  });

  return {
    signature: {
      algo: "SECP256K1",
      encoding: "base16",
      encodedValue: signature,
    },
    endpoint: {
      pallet: "statefulStorage",
      extrinsic: "applyItemActionsWithSignatureV2",
    },
    type: "itemActions",
    payload: payloadArguments,
  };
}

export async function createSignedLogInPayload(
  userAddress: string,
  signatureFn: SignatureFn,
  payloadArguments: CreateSignedLogInPayloadArguments,
): Promise<any> {
  const loginCaip122 = `text
    ${payloadArguments.domain} wants you to sign in with your Frequency account:${userAddress}
    
    URI: ${payloadArguments.uri}
    Version: ${payloadArguments.version}
    Nonce: ${payloadArguments.nonce}
    Chain ID: frequency:${payloadArguments.chainId}
    Issued At: ${payloadArguments.issuedAt}
  `;
  const signature = await signatureFn({
    method: "personal_sign",
    params: [userAddress, loginCaip122],
  });

  return {
    signature: {
      algo: "SECP256K1",
      encoding: "base16",
      encodedValue: signature,
    },
    type: "login",
    payload: {
      message: loginCaip122,
    },
  };
}
