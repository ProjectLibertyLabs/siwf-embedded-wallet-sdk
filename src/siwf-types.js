// TODO: Use the siwf library instead of copying these types?
export {};
