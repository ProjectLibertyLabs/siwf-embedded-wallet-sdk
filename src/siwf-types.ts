// TODO: Use the siwf library instead of copying these types?
// NOTE: The `siwf` types do not support 'Secp256k1' for `AlgorithmType`

import { HexString } from "@frequency-chain/ethereum-utils";

type CurveType = "Sr25519" | "Secp256k1";

type AlgorithmType = "SR25519" | "SECP256K1";

type SupportedPayload = unknown;

export type SignedPayload = Uint8Array | SupportedPayload;

export interface SiwfPublicKey {
  encodedValue: string;
  encoding: "base58" | "base16";
  format: "ss58" | "eip-55";
  type: CurveType;
}

interface SiwfResponsePayloadEndpoint {
  pallet: string;
  extrinsic: string;
}

interface SiwfResponsePayloadBase {
  signature: {
    algo: AlgorithmType;
    encoding: "base16";
    encodedValue: string;
  };
  endpoint?: SiwfResponsePayloadEndpoint;
  type: string;
  payload: Record<string, unknown>;
}

export interface AddProviderPayloadArguments extends Record<string, unknown> {
  authorizedMsaId: string;
  schemaIds: number[];
  expiration: number;
}

export interface SiwfResponsePayloadAddProvider
  extends SiwfResponsePayloadBase {
  endpoint: {
    pallet: "msa";
    extrinsic: "createSponsoredAccountWithDelegation" | "grantDelegation";
  };
  type: "addProvider";
  payload: AddProviderPayloadArguments;
}

export interface ItemActionsPayloadArguments extends Record<string, unknown> {
  schemaId: number;
  targetHash: number;
  expiration: number;
  actions: {
    type: "addItem";
    payloadHex: HexString;
  }[];
}

export interface SiwfResponsePayloadItemActions
  extends SiwfResponsePayloadBase {
  endpoint: {
    pallet: "statefulStorage";
    extrinsic: "applyItemActionsWithSignatureV2";
  };
  type: "itemActions";
  payload: ItemActionsPayloadArguments;
}

export interface ClaimHandlePayloadArguments extends Record<string, unknown> {
  baseHandle: string;
  expiration: number;
}

export interface SiwfResponsePayloadClaimHandle
  extends SiwfResponsePayloadBase {
  endpoint: {
    pallet: "handles";
    extrinsic: "claimHandle";
  };
  type: "claimHandle";
  payload: {
    baseHandle: string;
    expiration: number;
  };
}

export type SiwfResponsePayload =
  | SiwfResponsePayloadAddProvider
  | SiwfResponsePayloadItemActions
  | SiwfResponsePayloadClaimHandle
  | SiwfResponsePayloadBase;

export interface CreateSignedLogInPayloadArguments
  extends Record<string, unknown> {
  domain: string;
  uri: string;
  version: string;
  chainId: string;
  nonce?: string;
  issuedAt?: string;
}

interface SiwfResponseCredentialBase {
  "@context": [
    "https://www.w3.org/ns/credentials/v2",
    "https://www.w3.org/ns/credentials/undefined-terms/v2",
  ];
  type: string[];
  issuer: string;
  validFrom?: string;
  credentialSchema: {
    type: "JsonSchema";
    id: string;
  };
  credentialSubject: Record<string, unknown>;
  proof: {
    type: "DataIntegrityProof";
    created?: string;
    verificationMethod: string;
    cryptosuite: "eddsa-rdfc-2022";
    proofPurpose: "assertionMethod";
    proofValue: string;
    expirationDate?: string;
    validUntil?: string;
  };
}

export interface SiwfResponseCredentialEmail
  extends SiwfResponseCredentialBase {
  type: ["VerifiedEmailAddressCredential", "VerifiableCredential"];
  credentialSubject: {
    id: string;
    emailAddress: string;
    lastVerified: string;
  };
}

export interface SiwfResponseCredentialPhone
  extends SiwfResponseCredentialBase {
  type: ["VerifiedPhoneNumberCredential", "VerifiableCredential"];
  credentialSubject: {
    id: string;
    phoneNumber: string;
    lastVerified: string;
  };
}

export interface SiwfResponseCredentialGraph
  extends SiwfResponseCredentialBase {
  type: ["VerifiedGraphKeyCredential", "VerifiableCredential"];
  credentialSubject: {
    id: string;
    encodedPublicKeyValue: string;
    encodedPrivateKeyValue: string;
    encoding: "base16";
    format: "bare";
    type: "X25519";
    keyType: "dsnp.public-key-key-agreement";
  };
}

export type SiwfResponseCredential =
  | SiwfResponseCredentialEmail
  | SiwfResponseCredentialPhone
  | SiwfResponseCredentialGraph;

export interface SiwfResponse {
  userPublicKey: SiwfPublicKey;
  payloads: SiwfResponsePayload[];
  credentials?: SiwfResponseCredential[];
}
