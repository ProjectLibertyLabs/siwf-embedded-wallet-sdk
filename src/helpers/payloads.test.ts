import { describe, it, expect } from "vitest";
import { createSignedAddProviderPayload } from "./payloads";
import { SiwfResponsePayloadAddProvider } from "../siwf-types";


describe("createSignedAddProviderPayload", () => {
  it("returns the correct payload", async () => {
    const userAddress = "0x1234"
    const signatureFn = async (_request) => "fake-signature"

    const payload: SiwfResponsePayloadAddProvider = await createSignedAddProviderPayload(
      userAddress,
      signatureFn,
      {
        authorizedMsaId: 1n,
        schemaIds: [8, 9, 10, 15],
        expiration: 100,
      },
    )

    expect(payload).toMatchSnapshot()
  });
});
