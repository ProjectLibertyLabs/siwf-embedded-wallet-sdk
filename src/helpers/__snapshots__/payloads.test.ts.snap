// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createSignedAddProviderPayload > returns the correct payload 1`] = `
{
  "endpoint": {
    "extrinsic": "createSponsoredAccountWithDelegation",
    "pallet": "msa",
  },
  "payload": {
    "authorizedMsaId": 1,
    "expiration": 100,
    "schemaIds": [
      8,
      9,
      10,
      15,
    ],
  },
  "signature": {
    "algo": "SECP256K1",
    "encodedValue": "0x584322b9b1d988ba4ae457996a28cfcf3721ba3e7060aa8ddf315924234985fd14e1b84f0ab538c9eb36481e7cfd412fbcd7600831c405dbef361e87afbe94251b",
    "encoding": "base16",
  },
  "type": "addProvider",
}
`;

exports[`createSignedClaimHandlePayload > returns the correct payload 1`] = `
{
  "endpoint": {
    "extrinsic": "claimHandle",
    "pallet": "handles",
  },
  "payload": {
    "baseHandle": "Alice",
    "expiration": 100,
  },
  "signature": {
    "algo": "SECP256K1",
    "encodedValue": "0x6c6a34349c21810fa848f48d2ab97136b668c4643ef1bc9a2a46a538f42c4dd861570926653265c439026ecf275c61909a8bae356c1892cd1701cbf650b5b11c1b",
    "encoding": "base16",
  },
  "type": "claimHandle",
}
`;

exports[`createSignedGraphKeyPayload > returns the correct payload 1`] = `
{
  "endpoint": {
    "extrinsic": "applyItemActionsWithSignatureV2",
    "pallet": "statefulStorage",
  },
  "payload": {
    "actions": [
      {
        "payloadHex": "0x40a6836ea489047852d3f0297f8fe8ad6779793af4e9c6274c230c207b9b825026",
        "type": "addItem",
      },
    ],
    "expiration": 100,
    "schemaId": 7,
    "targetHash": 0,
  },
  "signature": {
    "algo": "SECP256K1",
    "encodedValue": "0xa90f53294466f6531e7e5e3a833497e613270ac97d0eb3ec1bff25bebc2ca6ff1cef743402012a5fbd1488ce1f6e077dc760721be8f9d46a74d9a5f30986696a1c",
    "encoding": "base16",
  },
  "type": "itemActions",
}
`;
